apollo-gradle-plugin-incubating/src/main/kotlin/com/apollographql/apollo/gradle/api/Service.kt

```diff
@@ -1,60 +1,23 @@
 package com.apollographql.apollo.gradle.api
 
-import com.apollographql.apollo.gradle.api.Introspection
+import com.apollographql.apollo.gradle.internal.DefaultIntrospection
 import groovy.lang.Closure
 import org.gradle.api.Action
+import org.gradle.api.provider.ListProperty
+import org.gradle.api.provider.Property
 
-class Service(val name: String): CompilerParams by DefaultCompilerParams() {
-  /**
-   * Place where the schema.json file is.
-   * This path is relative to the current project directory.
-   * schemaFilePath can point outside of src/{foo}/graphql but in that case, you'll want
-   * to define rootPackageName else fragments/types will be stored at the root of the namespace.
-   * By default, the plugin will look for a schema.json file in src/{foo}/graphql
-   */
-  var schemaFilePath: String? = null
+interface Service: CompilerParams {
+  fun introspection(configure: Action<in Introspection>)
 
-  /**
-   * Place where the graphql files are searched.
-   * This path is relative to the current sourceSet (e.g src/{foo}/graphql/{sourceFolderPath})
-   * By default, this is the directory where the schema.json is stored or "." if the schema is outside.
-   * You need to define this if you have two or more services whose schema is stored outside src/{foo}/graphql.
-   * If not, you can certainly omit it.
-   */
-  var sourceFolderPath: String? = null
+  val schemaPath: Property<String>
+  fun schemaPath(schemaFilePath: String)
 
-  var rootPackageName: String? = null
+  val sourceFolder: Property<String>
```

Comment:
The thing is android variants... `sourceFolder` has to be a string because the plugin will look into:

* src/main/graphql/$sourceFolder
* src/debug/graphql/$sourceFolder
* src/demoDebug/graphql/$sourceFolder

Service is really DSL oriented, it's just a helper to configure the tasks. If the user wants to do more configuration, he/she will have to use the `CompilationUnit`.

Reply:
Thinking about this again, I think it would be good to support Android Variants in more explicit way and keep these as files. 

This `service` idea can work pretty well with variants. Imagine you have a service and you configure common parts first (just like in `defaultConfig` in Android plugin). And then you can do per-variant configuration that would override the common configuration.

Reply:
Something like below?

```groovy
apollo {
  service("github") {
    // will only look in src/main/graphql/github by default
    sourceFolder = file("github") 

    productFlavors {
      demo {
        // will only look in src/main/graphql/github-demo for the demo flavor
        sourceFolder = file("github-demo") 
      }
    }
  }
}
```

I don't like this too much as this is against the android standard sourceSets resolution (https://developer.android.com/studio/build/build-variants#sourceset-build). In the example above, files in `src/debug/graphql/github` would never be picked up.

Reply:
We might want to block https://github.com/apollographql/apollo-android/pull/1673 until we decide something there. Once the plugin is promoted, it'll be much more difficult to do impactful changes.